### PR TITLE
P1686R2 Mandating the Standard Library: Clause 27 - Time library

### LINKED ISSUE #3428: [2019-11 LWG Motion 26] P1686R2 Mandating the Standard Library: Clause 27 - Time library

### PR DESCRIPTION
**Pull request by zygoloid**
Fixes #3428.
Fixes cplusplus/papers#463.

### CODE REVIEW

File: source/time.tex [hunk 1]
```diff
@@ -1400,8 +1392,8 @@
 
 \begin{itemdescr}
 \pnum
-\returns
-\tcode{duration(rep_++)}.
+\effects
+Equivalent to: \tcode{return duration(rep_++)};
```
Comment (johelegp) [inline source/time.tex]:
```suggestion
Equivalent to: \tcode{return duration(rep_++);}
```

File: source/time.tex [hunk 2]
```diff
@@ -1426,8 +1418,8 @@
 
 \begin{itemdescr}
 \pnum
-\returns
-\tcode{duration(rep_-{}-)}.
+\effects
+Equivalent to: \tcode{return duration(rep_-{}-)};
```
Comment (johelegp) [inline source/time.tex]:
```suggestion
Equivalent to: \tcode{return duration(rep_-{}-);}
```

File: source/time.tex [hunk 3]
```diff
@@ -1615,9 +1606,8 @@
 
 \begin{itemdescr}
 \pnum
-\remarks
-This operator shall not participate in overload
-resolution unless \tcode{Rep1} is implicitly convertible to \tcode{common_type_t<Rep1, Rep2>}.
+\constraints
+\tcode{is_convertible_v<Rep1, common_type_t<Rep1, Rep2>> } is \tcode{true}.
```
Comment (johelegp) [inline source/time.tex]:
```suggestion
\tcode{is_convertible_v<Rep1, common_type_t<Rep1, Rep2>>} is \tcode{true}.
```

File: source/time.tex [hunk 4]
```diff
@@ -1633,9 +1623,8 @@
 
 \begin{itemdescr}
 \pnum
-\remarks
-This operator shall not participate in overload
-resolution unless \tcode{Rep2} is implicitly convertible to \tcode{common_type_t<Rep1, Rep2>}
+\constraints
+is_convertible_v<Rep2, common_type_t<Rep1, Rep2>> is \tcode{true}
```
Comment (johelegp) [inline source/time.tex]:
```suggestion
\tcode{is_convertible_v<Rep2, common_type_t<Rep1, Rep2>>} is \tcode{true}
```

File: source/time.tex [hunk 5]
```diff
@@ -1669,9 +1658,8 @@
 
 \begin{itemdescr}
 \pnum
-\remarks
-This operator shall not participate in overload
-resolution unless \tcode{Rep2} is implicitly convertible to \tcode{common_type_t<Rep1, Rep2>} and
+\constraints
+is_convertible_v<Rep2, common_type_t<Rep1, Rep2>> is \tcode{true} and
```
Comment (johelegp) [inline source/time.tex]:
```suggestion
\tcode{is_convertible_v<Rep2, common_type_t<Rep1, Rep2>>} is \tcode{true} and
```

File: source/time.tex [hunk 6]
```diff
@@ -2252,8 +2231,8 @@
 
 \begin{itemdescr}
 \pnum
-\returns
-\tcode{time_point\{d_++\}}.
+\effects
+Equivalent to: \tcode{return time_point\{d_++}\};
```
Comment (johelegp) [inline source/time.tex]:
```suggestion
Equivalent to: \tcode{return time_point\{d_++\};}
```

File: source/time.tex [hunk 7]
```diff
@@ -2278,8 +2257,8 @@
 
 \begin{itemdescr}
 \pnum
-\returns
-\tcode{time_point\{d_--\}}.
+\effects
+Equivalent to: \tcode{return {time_point\{d_-{}-\}}};
```
Comment (johelegp) [inline source/time.tex]:
```suggestion
Equivalent to: \tcode{return time_point\{d_-{}-\};}
```

File: source/time.tex [hunk 8]
```diff
@@ -2605,8 +2580,7 @@
 
 \begin{itemdescr}
 \pnum
-\requires \tcode{system_clock::duration::min() < system_clock::duration::zero()} shall
-be \tcode{true}.\\
+\tcode{system_clock::duration::min() < system_clock::duration::zero()} is \tcode{true}.
```
Comment (johelegp) [inline source/time.tex]:
```suggestion
\constraints
\tcode{system_clock::duration::min() < system_clock::duration::zero()} is \tcode{true}.
```

File: source/time.tex [hunk 9]
```diff
@@ -2655,10 +2629,8 @@
 
 \begin{itemdescr}
 \pnum
-\remarks
-This operator shall not participate in overload resolution if
-\tcode{treat_as_floating_point_v<typename Duration::rep>} is \tcode{true},
-or if \tcode{Duration\{1\} >= days\{1\}}.
+\constraints
+\tcode{treat_as_floating_point_v<typename Duration::rep>} is false, and \tcode{Duration\{1\}} \tcode{<} \tcode{days\{1\}}.
```
Comment (johelegp) [inline source/time.tex]:
```suggestion
\tcode{treat_as_floating_point_v<typename Duration::rep>} is \tcode{false}, and
\tcode{Duration\{1\} < days\{1\}} is \tcode{true}.
```

File: source/time.tex [hunk 10]
```diff
@@ -3662,13 +3634,13 @@
...
 \tcode{SourceClock::to_sys(t)} is well-formed.
-If \tcode{SourceClock::to_sys(t)}
-does not return \tcode{sys_time<Duration>},
-where \tcode{Duration} is a valid \tcode{chrono::duration} specialization,
-the program is ill-formed.
+
+\pnum
+\mandates
+\tcode{SourceClock::to_sys(t)} returns a \tcode{sys_time<Duration>},
+where \tcode{Duration} is a valid \tcode{chrono::duration} specialization.
```
Comment (johelegp) [inline source/time.tex]:
```suggestion
where \tcode{Duration} is a valid \tcode{duration} specialization.
```

File: source/time.tex [hunk 11]
```diff
@@ -3693,13 +3665,13 @@
...
 \tcode{DestClock::from_sys(t)} is well-formed.
-If \tcode{DestClock::from_sys(t)} does not return
-\tcode{time_point<DestClock, Duration>},
-where \tcode{Duration} is a valid \tcode{chrono::duration} specialization,
-the program is ill-formed.
+
+\pnum
+\mandates
+\tcode{DestClock::from_sys(t)} returns a \tcode{time_point<DestClock, Duration>},
+where \tcode{Duration} is a valid \tcode{chrono::duration} specialization.
```
Comment (johelegp) [inline source/time.tex]:
```suggestion
where \tcode{Duration} is a valid \tcode{duration} specialization.
```

File: source/time.tex [hunk 12]
```diff
@@ -3726,13 +3698,13 @@
...
 \tcode{SourceClock::to_utc(t)} is well-formed.
-If \tcode{SourceClock::to_utc(t)} does not return
-\tcode{utc_time<Duration>},
-where \tcode{Duration} is a valid \tcode{chrono::duration} specialization,
-the program is ill-formed.
+
+\pnum
+\mandates
+\tcode{SourceClock::to_utc(t)}  returns a \tcode{utc_time<Duration>},
+where \tcode{Duration} is a valid \tcode{chrono::duration} specialization.
```
Comment (johelegp) [inline source/time.tex]:
```suggestion
where \tcode{Duration} is a valid \tcode{duration} specialization.
```

File: source/time.tex [hunk 13]
```diff
@@ -3757,13 +3729,13 @@
...
+\constraints
 \tcode{DestClock::from_utc(t)} is well-formed.
-If \tcode{DestClock::from_utc(t)} does not return
-\tcode{time_point<DestClock, Duration>},
-where \tcode{Duration} is a valid \tcode{chrono::duration} specialization,
-the program is ill-formed.
+
+\pnum
+\mandates
+\tcode{DestClock::from_utc(t)}  returns a \tcode{time_point<DestClock, Duration>},
```
Comment (johelegp) [inline source/time.tex]:
```suggestion
\tcode{DestClock::from_utc(t)} returns a \tcode{time_point<DestClock, Duration>},
```

File: source/time.tex [hunk 14]
```diff
@@ -3757,13 +3729,13 @@
...
+where \tcode{Duration} is a valid \tcode{chrono::duration} specialization.
```
Comment (johelegp) [inline source/time.tex]:
```suggestion
where \tcode{Duration} is a valid \tcode{duration} specialization.
```

File: source/time.tex [hunk 15]
```diff
@@ -3662,13 +3634,13 @@
 
 \begin{itemdescr}
 \pnum
-\remarks
-This function does not participate in overload resolution unless
+\constraints
 \tcode{SourceClock::to_sys(t)} is well-formed.
```
Comment (johelegp) [inline source/time.tex]: I think I remember one such change from `\remarks` to `\constraints` also changed "well-formed" to "valid".

Comment (zygoloid) [reply source/time.tex]: "well-formed" is a defined term; "valid" is not. We should at least pick one and apply it consistently. Filed #3499 to track that.

File: source/time.tex (hunk 10 above)
Comment (zygoloid) [reply source/time.tex]: Hmm. [contents]p3 requires the `chrono::` here, but we're certainly not consistent about that.

**Comment by jensmaurer** [comment]
`time.tex:Overfull \hbox (1.13849pt too wide) in paragraph at lines 2582--2588`

File: source/time.tex (hunk 10 above)
Comment (zygoloid) [reply source/time.tex]: This is pre-existing, and the pre-existing wording is in line with [contents]p3. We should do something about this (we probably want an equivalent of [fs.req.namespace] for std::chrono), but not as part of applying this motion.

Comment (zygoloid) [reply source/time.tex]: Filed #3510.

**Pull request merged**